cluster-api/machines-api/types.go

```diff
@@ -0,0 +1,160 @@
+	// +optional
+	NodeRef *api.ObjectReference
+
+	// When was this status last observed
+	// +optional
+	LastUpdated metav1.Time
+
+	// Indicates whether or not the Machine is fully reconciled. When a
+	// controller observes that the spec has changed and no longer matches
+	// reality, it should update Ready to false before reconciling the
+	// state, and then set back to true when the state matches the spec.
+	Ready bool
+
+	// If set, indicates that there is a problem reconciling state, and
+	// will be set to a token value suitable for machine interpretation.
+	// +optional
+	ErrorReason *MachineStatusError
+
+	// +optional
+	// If set, indicates that there is a problem reconciling state, and
+	// will be set to a human readable string to indicate the problem.
+	ErrorMessage *string
+}
+
+type MachineStatusError string
+
+const (
+	// Represents that the combination of configuration in the MachineSpec
+	// is not supported by this cluster. This is not a transient error, but
+	// indicates a state that must be fixed before progress can be made.
+	//
+	// Example: the ProviderConfig specifies an instance type that doesn't exist,
+	InvalidConfigurationMachineError MachineStatusError = "InvalidConfiguration"
+
+	// This indicates that the MachineSpec has been updated in a way that
+	// is not supported for reconciliation on this cluster. The spec may be
+	// completely valid from a configuration standpoint, but the controller
+	// does not support changing the real world state to match the new
+	// spec.
+	//
+	// Example: the responsible controller is not capable of changing the
+	// container runtime from docker to rkt.
+	UnsupportedChangeMachineError MachineStatusError = "UnsupportedChange"
+
+	// This generally refers to exceeding one's quota in a cloud provider,
+	// or running out of physical machines in an on-premise environment.
+	InsufficientResourcesMachineError MachineStatusError = "InsufficientResources"
+
+	// There was an error while trying to create a Node to match this
+	// Machine. This may indicate a transient problem that will be fixed
+	// automatically with time, such as a service outage, or a terminal
+	// error during creation that doesn't match a more specific
+	// MachineStatusError value.
+	//
+	// Example: timeout trying to connect to GCE.
+	CreateMachineError MachineStatusError = "CreateError"
+
+	// An error was encountered while trying to delete the Node that this
+	// Machine represents. This could be a transient or terminal error, but
+	// will only be observable if the provider's Machine controller has
+	// added a finalizer to the object to more gracefully handle deletions.
+	//
+	// Example: cannot resolve EC2 IP address.
+	DeleteMachineError MachineStatusError = "DeleteError"
+)
+
+type MachineVersionInfo struct {
+	// Semantic version of kubelet to run
+	Kubelet string
+
+	// Name/version of container runtime
+	ContainerRuntime ContainerRuntimeInfo
+}
+
+type ContainerRuntimeInfo struct {
```

Comment:
What is the scope of this particular object/concept? Should we expect to see `{Rkt,Frakti,Containerd}RuntimeConfig` structs at some point?

Reply:
The intended purpose was to:

1) Know what runtime to install at provisioning time. It's completely fine for an implementation to say "sorry, I only know how to install Docker. Anything else will result in an `UnsupportedConfiguration` error." Or even "I have no idea how to install that version of Docker," etc.
2) Know how to set kubelet's `--container-runtime` flag.

You're right to question how this will evolve in the future, with cluster admins potentially wanting to fine-tune the settings of the runtime itself. I think it would largely follow the same pattern we use to handle the provider-specific configuration for `Machines`. At first, we would likely have an opaque blob to capture all of the settings that were fed to the runtime installer, which would allow each container runtime to version its configuration independently of the Machines API. Then, potentially upgrade this to an ObjectReference so that it identical configuration wouldn't have to be inlined with every object that uses it. Any time we have configuration that seems to be useful across every runtime we support, we can graduate it out of the opaque config blob and into the `ContainerRuntimeInfo` struct if desired.

I think the `ContainerRuntimeInfo` is a good candidate for culling in v1alpha1, actually, until there's a strong need to add it. Definitely worth discussion.